lib/types.nix

```diff
@@ -65,6 +65,10 @@ rec {
       # definition values and locations (e.g. [ { file = "/foo.nix";
       # value = 1; } { file = "/bar.nix"; value = 2 } ]).
       merge ? mergeDefaultOption
+    , # Whether this type has a value representing nothingness. If it does,
+      # this should be a value of the form { value = <the nothing value>; }
+      # If it doesn't, this should be {}
+      emptyValue ? {}
```

Comment:
I suppose a nested field is used to express the optional nature of `emptyValue`, which can already be expressed with an optional field.

```
args@{ emptyValue ? null }: {
  # ...
} // lib.optionalAttrs (args ? "emptyValue") {
  inherit emptyValue;
}
```

Although construction is a bit cumbersome, we remove an indirection and some ambiguity.

Reply:
Yeah this was how i wanted to implement it initially, but then things like propagation of this value would be annoying, e.g. in `unique` and `coercedTo`. This approach is also more uniform with all other parameters of `mkOptionType`.